File: packages/railtracks/src/railtracks/llm/tools/parameters/ref_parameter.py
from typing import Any, Dict, Optional

from ._base import Parameter


class RefParameter(Parameter):
    param_type: str = "object"  # referenced schemas are always 'object' type

    def __init__(
        self,
        name: str,
        ref_path: str,
        description: Optional[str] = None,
        required: bool = True,
        default: Any = None,
    ):
        super().__init__(name, description, required, default)
        self.ref_path = ref_path

    def to_json_schema(self) -> Dict[str, Any]:
        schema = {"$ref": self.ref_path}
        if self.description:
            schema["description"] = self.description

        if self.default is not None:
            schema["default"] = self.default

        if self.enum:
            schema["enum"] = self.enum

        return schema

    def __repr__(self) -> str:
        return (
            f"RefParameter(name={self.name!r}, ref_path={self.ref_path!r}, "
            f"description={self.description!r}, required={self.required!r}, default={self.default!r})"
        )

File: dotnet/src/Railtracks.Core/Llm/Tools/RefParameter.cs
namespace Railtracks.Llm.Tools;

/// <summary>Parameter representing a JSON-schema reference. Ported from <c>RefParameter</c>.</summary>
public sealed class RefParameter : Parameter
{
    public RefParameter(
        string name,
        string refPath,
        string? description = null,
        bool required = true,
        object? @default = null)
        : base(name, description, required, @default)
    {
        RefPath = refPath;
        ParamType = "object"; // referenced schemas are always object type
    }

    public string RefPath { get; }

    public override Dictionary<string, object?> ToJsonSchema()
    {
        var schema = new Dictionary<string, object?> { ["$ref"] = RefPath };

        if (!string.IsNullOrEmpty(Description))
            schema["description"] = Description;
        if (Default is not null)
            schema["default"] = Default;
        if (Enum is { Count: > 0 })
            schema["enum"] = Enum;

        return schema;
    }

    public override string ToString() =>
        $"RefParameter(name='{Name}', ref_path='{RefPath}', description='{Description}', " +
        $"required={Required}, default={Default})";
}
